concordium-consensus/src/Concordium/KonsensusV1/Consensus/CatchUp.hs

```diff
@@ -13,3 +54,490 @@ module Concordium.KonsensusV1.Consensus.CatchUp where
+                | cusCurrentRound < _rsCurrentRound rs =
+                    rtTimeoutCertificate <$> _rsPreviousRoundTimeout rs
+                | otherwise = Absent
+        cutdCurrentRoundQuorumMessages <- getQuorumMessages
+        cutdCurrentRoundTimeoutMessages <- getTimeoutMessages
+        return $ CatchUpTerminalData{..}
+    highQC = skovData ^. roundStatus . rsHighestCertifiedBlock . to cbQuorumCertificate
+    getQCs
+        | qcRound highQC < cusLastFinalizedRound = return []
+        | qcRound highQC == blockRound lfBlock || blockRound lfBlock <= cusLastFinalizedRound =
+            return [highQC]
+        | otherwise = do
+            LowLevel.lookupLatestFinalizationEntry <&> \case
+                Nothing -> [highQC]
+                Just finEntry -> [feSuccessorQuorumCertificate finEntry, highQC]
+    ourCurrentRound = skovData ^. roundStatus . rsCurrentRound
+    getQuorumMessages
+        | cusCurrentRound == ourCurrentRound = do
+            let qms = skovData ^.. currentQuorumMessages . smFinalizerToQuorumMessage . traversed
+            return $! filter newToPeer qms
+        | cusCurrentRound < ourCurrentRound = do
+            let qms = skovData ^.. currentQuorumMessages . smFinalizerToQuorumMessage . traversed
+            return qms
+        | otherwise = do
+            return []
+      where
+        newToPeer qm = case Map.lookup (qmBlock qm) cusCurrentRoundQuorum of
+            Nothing -> True
+            Just s -> not (memberFinalizerSet (qmFinalizerIndex qm) s)
+    getTimeoutMessages
+        | cusCurrentRound <= ourCurrentRound =
+            case skovData ^. currentTimeoutMessages of
+                Absent -> return []
+                Present TimeoutMessages{..} -> do
+                    let (filterFirst, filterSecond) = case cusCurrentRoundTimeouts of
+                            Present TimeoutSet{..}
+                                | cusCurrentRound < ourCurrentRound -> (id, id)
+                                | tsFirstEpoch + 1 == tmFirstEpoch -> (filter2, id)
+                                | tsFirstEpoch == tmFirstEpoch -> (filter1, filter2)
+                                | tsFirstEpoch == tmFirstEpoch + 1 -> (id, filter1)
+                              where
+                                filterFun ts tm =
+                                    not $ memberFinalizerSet (tmFinalizerIndex $ tmBody tm) ts
+                                filter1 = filter (filterFun tsFirstEpochTimeouts)
+                                filter2 = filter (filterFun tsSecondEpochTimeouts)
+                            _ -> (id, id)
+                    return $
+                        filterFirst (Map.elems tmFirstEpochTimeouts)
+                            <> filterSecond (Map.elems tmSecondEpochTimeouts)
+        | otherwise =
+            return []
+
+data TerminalDataResult
+    = TerminalDataResultValid {tdrProgress :: Bool}
+    | TerminalDataResultInvalid {tdrProgress :: Bool}
```

Comment:
I'm struggling to understand this type, and the meaning of `tdrProgress`. Can you please add documentation explaining it.